### PR TITLE
upload translation for chapter12

### PR DESCRIPTION
**Pull request by Zheng7426**
@yukie1412 Hi Yukie, 

Please check the translation for chapter 12 when you have time. Thanks. 

Sincerely,
Zheng

### CODE REVIEW

File: 12_language.md [hunk 1]
```diff
@@ -0,0 +1,670 @@
...
+
+求值器可以决定一门编程语言中表达式的含义，而它本身也只是一个程序。
+
+quote}}
+
+{{index "Abelson, Hal", "Sussman, Gerald", SICP, "project chapter"}}
+
+{{figure {url: "img/chapter_picture_12.jpg", alt: "Picture of an egg with smaller eggs inside", chapter: "framed"}}}
+
+构建你自己的((编程语言))其实出乎意料地简单(只要你的要求不是太高)而且很启发人。
```
Comment (ghost) [inline 12_language.md]: the bracket around 只要你的要求不是太高 should be `（）` instead (中文输入的 bracket)

File: 12_language.md [hunk 2]
```diff
@@ -0,0 +1,670 @@
...
+  program = skipSpace(program);
+  let match, expr;
+  if (match = /^"([^"]*)"/.exec(program)) {
+    expr = {type: "value", value: match[1]};
+  } else if (match = /^\d+\b/.exec(program)) {
+    expr = {type: "value", value: Number(match[0])};
+  } else if (match = /^[^\s(),#"]+/.exec(program)) {
+    expr = {type: "word", name: match[0]};
+  } else {
+    throw new SyntaxError("Unexpected syntax: " + program);
```
Comment (ghost) [inline 12_language.md]: per usual, should translate

File: 12_language.md [hunk 3]
```diff
@@ -0,0 +1,670 @@
...
+//    operator: {type: "word", name: "+"},
+//    args: [{type: "word", name: "a"},
+//           {type: "value", value: 10}]}
+```
+
+{{index "error message"}}
+
+这行得通！它在出故障的时候并不会给我们非常有用的信息，也不保存每一个表达式起始的行和列(可能在之后报告错误的时候会有帮助)，但是它已经足够满足我们的目的了。
+
+## 求值器
```
Comment (yukie1412) [inline 12_language.md:192]: // todo

File: 12_language.md (hunk 1 above)
Comment (Zheng7426) [reply 12_language.md]: Ok.

File: 12_language.md [hunk 4]
```diff
@@ -0,0 +1,670 @@
...
+
+{{index [function, application]}}
+
+应用(Applications)则更复杂一些。如果它们属于像 `if` 那样的((特殊形式))，我们不对其进行求值，而是将参数表达式与作用域一起传递给处理这个形式的函数。如果应用是普通调用，那我们对运算符进行求值，确认其是否是函数，并且以求值后的参数来调用它。
+
+为了表示 Egg 的函数值，我们使用普通的 JavaScript 函数。我们 [之后](language#egg_fun) 在定义了名为 `fun` 的特殊形式的时候，会回到这个话题。
+
+{{index readability, "evaluate function", recursion, parsing}}
+
+`evaluate` 的递归结构与解析器的结构类似，而且两者都反映出该语言本身的结构。将解析器和求值器整合于一体、并在解析的时候进行求值也是可能的，但是将它们如此分离使得该程序更容易被理解。
```
Comment (ghost) [inline 12_language.md]: 但是 sounds weird here, how about 不过？
also i think 可以的 sounds better than 可能的，thoughts?

File: 12_language.md [hunk 5]
```diff
@@ -0,0 +1,670 @@
...
+    return evaluate(args[1], scope);
+  } else {
+    return evaluate(args[2], scope);
+  }
+};
+```
+
+{{index "conditional execution", "ternary operator", "?: operator", "conditional operator"}}
+
+Egg 的 `if` 构造语句必须接受三个参数，它会首先求解第一个参数，且在其结果不为 `false` 值的情况下求解第二个参数。否则的话，第三个参数会被求解。与 JavaScript 的 `if` 相比，这个 Egg 中的 `if` 形式与 JavaScript 的三元 `?:` 操作符更为相似。它是一个表达式，而不是语句，而且它产生一个值，也就是第二个或第三个参数的结果。
```
Comment (ghost) [inline 12_language.md]: i think 值 is redundant here, "且在其结果不为 `false`" seems sufficient to me

File: 12_language.md [hunk 6]
```diff
@@ -0,0 +1,670 @@
...
+## 特殊形式
+
+{{index "special form", "specialForms object"}}
+
+`specialForms` 对象用于定义 Egg 中的特殊句法，它将单词和求解这种形式的函数联系了起来。当前该对象是空的，我们加上 `if`。
+
+```{includeCode: true}
+specialForms.if = (args, scope) => {
+  if (args.length != 3) {
+    throw new SyntaxError("Wrong number of args to if");
```
Comment (ghost) [inline 12_language.md]: Q: should we localize syntax errors?

File: 12_language.md [hunk 7]
```diff
@@ -0,0 +1,670 @@
...
+
+```{test: no}
+Object.prototype.hasOwnProperty.call(scope, name);
+```
+
+{{if interactive
+
+```{test: no}
+specialForms.set = (args, scope) => {
+  // Your code here.
```
Comment (ghost) [inline 12_language.md]: localize

File: 12_language.md (hunk 2 above)
Comment (Zheng7426) [reply 12_language.md]: Sure.

File: 12_language.md (hunk 4 above)
Comment (Zheng7426) [reply 12_language.md]: Ok 👍

File: 12_language.md (hunk 5 above)
Comment (Zheng7426) [reply 12_language.md]: Good idea.

File: 12_language.md (hunk 6 above)
Comment (Zheng7426) [reply 12_language.md]: Sure thing.

**Pull request merged**